### PR TITLE
Add out variants for softmax and log_softmax

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75833

### CODE REVIEW

File: aten/src/ATen/native/SoftMax.cpp [hunk 1]
```diff
@@ -418,6 +418,11 @@ Tensor softmax(const Tensor& input_, const int64_t dim_, c10::optional<ScalarTyp
   return result;
 }
 
+Tensor& softmax_out(const Tensor& input_, const int64_t dim_, c10::optional<ScalarType> dtype, Tensor& output_) {
+  Tensor converted = dtype.has_value() ? input_.toType(dtype.value()) : input_;
```
Comment (ngimel) [inline aten/src/ATen/native/SoftMax.cpp:422]: should you have the same logic to handle fp16->fp32 conversion on cuda that regular softmax/logsoftmax has?

File: aten/src/ATen/native/SoftMax.cpp [hunk 2]
```diff
@@ -446,6 +451,11 @@ Tensor log_softmax(const Tensor& input_, const int64_t dim_, c10::optional<Scala
   return result;
 }
 
+Tensor& log_softmax_out(const Tensor& input_, const int64_t dim_, c10::optional<ScalarType> dtype, Tensor& output_) {
+  Tensor converted = dtype.has_value() ? input_.toType(dtype.value()) : input_;
+  return at::_log_softmax_out(output_, converted, dim_, false);
```
Comment (bdhirsh) [inline aten/src/ATen/native/SoftMax.cpp:456]: is there any reason we can't just promote these private API's to be public, instead of leaving them?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -4121,6 +4124,9 @@
 - func: softmax.int(Tensor self, int dim, ScalarType? dtype=None) -> Tensor
   variants: function, method
 
+- func: softmax.int_out(Tensor self, int dim, ScalarType? dtype=None, *, Tensor(a!) out) -> Tensor(a!)
+  variants: function
```
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:4128]: These out= ops should probably be marked `CompositeExplicitAutograd`, since out= ops aren't supported by autograd (and the "default" in native_functions.yaml is `CompositeImplicitAutograd`, which implies autograd support).

cc @albanD is any other action needed (like explicitly specifying the lack of autograd support)? Or will autograd automatically generate a nice error message when you try to use one of these ops with it.

File: aten/src/ATen/native/SoftMax.cpp (hunk 2 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/native/SoftMax.cpp:456]: I believe this is more of a short term fix to unblock our new edge runtime, but I think consolidating log_softmax with _log_softmax is probably a better solution. cc @gmagogsfm  to confirm

File: aten/src/ATen/native/SoftMax.cpp (hunk 1 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/native/SoftMax.cpp:422]: I am not familiar with CUDA. Do you know anyone who I can talk to?

Comment (tugsbayasgalan) [reply aten/src/ATen/native/SoftMax.cpp:422]: Ahh okay i think i know what you mean

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:4128]: If it is written in a "differentiable" way, then nothing to be done. If you call native ops, then yes you can make it CompositeExplicitAutograd to get a default error kernel.

File: aten/src/ATen/native/SoftMax.cpp [hunk 4]
```diff
@@ -446,6 +483,43 @@ Tensor log_softmax(const Tensor& input_, const int64_t dim_, c10::optional<Scala
...
+  Tensor output_temp;
+  if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half &&
+      dtype == ScalarType::Float) {
+    if (!output_.is_contiguous()) {
+      auto options =
+          TensorOptions().dtype(output_.dtype()).device(output_.device());
+      output_temp = at::empty(output_.sizes(), options);
+      at::_log_softmax_out(output_temp, input_, dim_, true);
+    } else {
+      at::_log_softmax_out(output_, input_, dim_, true);
```
Comment (ngimel) [inline aten/src/ATen/native/SoftMax.cpp:500]: You should check that `output_` is correct size before sending it? I don't think _log_softmax_out is doing it itself.

**Review by ngimel — APPROVED** [review]
Left a small comment, looks good. 
I agree with @bdhirsh that longer term we should make private out ops public, but for now to unblock these wrappers are fine, and they don't regress perf of existing paths.

File: aten/src/ATen/native/SoftMax.cpp [hunk 5]
```diff
@@ -438,6 +438,43 @@ Tensor softmax(const Tensor& input_, const int64_t dim_, c10::optional<ScalarTyp
...
+      auto options =
+          TensorOptions().dtype(output_.dtype()).device(output_.device());
+      output_temp = at::empty(output_.sizes(), options);
+      at::_softmax_out(output_temp, converted, dim_, false);
+    } else {
+      at::_softmax_out(output_, converted, dim_, false);
+    }
+  }
+
+  if (!output_.is_contiguous()) {
```
Comment (mruberry) [inline aten/src/ATen/native/SoftMax.cpp:470]: I don't think this is correct? Take a look at resize_output and https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch

Comment (tugsbayasgalan) [reply aten/src/ATen/native/SoftMax.cpp:470]: If the output tensor is not contiguous, we do this to work around because _softmax only works for contiguous tensors
